code/modules/atmospherics/gasmixtures/reactions.dm

```diff
@@ -770,19 +790,636 @@ nobiliumsuppression = INFINITY
+		return NO_REACTION
+	cached_gases[/datum/gas/hexane][MOLES] -= heat_efficency * 0.2
+	cached_gases[/datum/gas/plasma][MOLES] -= heat_efficency * 0.5
+	cached_gases[/datum/gas/carbon_dioxide][MOLES] += heat_efficency * 0.4
+
+	if(energy_used)
+		var/new_heat_capacity = air.heat_capacity()
+		if(new_heat_capacity > MINIMUM_HEAT_CAPACITY)
+			air.temperature = max(((temperature * old_heat_capacity + energy_used) / new_heat_capacity), TCMB)
+	return REACTING
+
+/datum/gas_reaction/hexane_n2o_filtering
+	priority = 10
+	name = "Hexane n2o filtering"
+	id = "hexane_n2o_filtering"
+
+/datum/gas_reaction/hexane_n2o_filtering/init_reqs()
+	min_requirements = list(
+		/datum/gas/hexane = MINIMUM_MOLE_COUNT,
+		/datum/gas/nitrous_oxide = MINIMUM_MOLE_COUNT,
+	)
+
+/datum/gas_reaction/hexane_n2o_filtering/react(datum/gas_mixture/air, datum/holder)
+	var/list/cached_gases = air.gases
+	var/temperature = air.temperature
+	var/old_heat_capacity = air.heat_capacity()
+	var/heat_efficency = min(temperature * 0.01, cached_gases[/datum/gas/hexane][MOLES], cached_gases[/datum/gas/nitrous_oxide][MOLES])
+	var/energy_used = heat_efficency * 100
+	ASSERT_GAS(/datum/gas/oxygen, air)
+	ASSERT_GAS(/datum/gas/nitrogen, air)
+	if ((cached_gases[/datum/gas/hexane][MOLES] - heat_efficency * 0.2 < 0 ) || (cached_gases[/datum/gas/nitrous_oxide][MOLES] - heat_efficency * 0.5 < 0)) //Shouldn't produce gas from nothing.
+		return NO_REACTION
+	cached_gases[/datum/gas/hexane][MOLES] -= heat_efficency * 0.2
+	cached_gases[/datum/gas/nitrous_oxide][MOLES] -= heat_efficency * 0.6
+	cached_gases[/datum/gas/oxygen][MOLES] += heat_efficency * 0.2
+	cached_gases[/datum/gas/nitrogen][MOLES] += heat_efficency * 0.6
+
+	if(energy_used)
+		var/new_heat_capacity = air.heat_capacity()
+		if(new_heat_capacity > MINIMUM_HEAT_CAPACITY)
+			air.temperature = max(((temperature * old_heat_capacity + energy_used) / new_heat_capacity), TCMB)
+	return REACTING
+
+/datum/gas_reaction/cyrion_b_decomp
+	priority = 11
+	name = "Cyrion B decomposition"
+	id = "cyrion_b_decomp"
+
+/datum/gas_reaction/cyrion_b_decomp/init_reqs()
+	min_requirements = list(
+		/datum/gas/nitrogen = MINIMUM_MOLE_COUNT,
+		/datum/gas/cyrion_b = MINIMUM_MOLE_COUNT
+	)
+
+/datum/gas_reaction/cyrion_b_decomp/react(datum/gas_mixture/air, datum/holder)
+	var/energy_released = 0
+	var/old_heat_capacity = air.heat_capacity()
+	var/list/cached_gases = air.gases //this speeds things up because accessing datum vars is slow
+	var/temperature = air.temperature
+	var/burned_fuel = 0
+	burned_fuel = min(20, cached_gases[/datum/gas/nitrogen][MOLES], cached_gases[/datum/gas/cyrion_b][MOLES])
+	if(cached_gases[/datum/gas/cyrion_b][MOLES] - burned_fuel < 0)
+		return NO_REACTION
+	cached_gases[/datum/gas/cyrion_b][MOLES] -= burned_fuel
+
+	if(burned_fuel)
+		energy_released += (250 * burned_fuel)
+
+		ASSERT_GAS(/datum/gas/oxygen, air)
+		cached_gases[/datum/gas/oxygen][MOLES] += burned_fuel * 0.3
+		ASSERT_GAS(/datum/gas/nitrogen, air)
```

Comment:
For cleanliness sake do all your asserting before the add/subtract. It might be very very slightly slower due to the list lookup time, but it's alright.